File: typescript/src/mesa.ts
import type { WhoamiResponse } from '@mesadev/rest';
import { normalizeAccessTokenAuthors, signAutomaticPrivateKeyAccessToken, type Author } from './api/access-token.js';
import { createRestClient, type RestClient } from './api/client.js';
import { type ApiResources, createApiResources } from './api/resources.js';
import { looksLikePrivateKey } from './api/credentials.js';
import { parsePrivateKey, type PrivateKeyCredential } from './api/private-key.js';
import { type Layout, type Repo } from './fs/layout.js';
import { MesaFileSystem, type MesaFileSystemConfig, type TelemetryConfig } from './fs/mesa-file-system.js';
import { InvalidApiUrlError, InvalidOptionsError, MissingPrivateKeyError, OrgResolutionError } from './lib/errors.js';

const DEFAULT_API_URL = 'https://api.mesa.dev/v1';
const PRIVATE_KEY_ENV_VAR = 'MESA_PRIVATE_KEY';

function getEnvVar(name: string): string | undefined {
  if (typeof process === 'undefined') {
    return undefined;
  }

  return process.env[name]?.trim();
}

function normalizeUrl(input: string): string {
  if (looksLikePrivateKey(input)) {
    throw new InvalidApiUrlError('Mesa private keys cannot be used as URLs');
  }

  let parsed: URL;
  try {
    parsed = new URL(input);
  } catch {
    throw new InvalidApiUrlError(input);
  }

  if (parsed.protocol !== 'https:' && parsed.protocol !== 'http:') {
    throw new InvalidApiUrlError(input);
  }

  parsed.pathname = parsed.pathname.replace(/\/+$/, '') || '/';
  return parsed.toString().replace(/\/$/, '');
}

/** Resolve an explicit private key or the environment fallback. */
function resolvePrivateKey(privateKey: string | undefined): PrivateKeyCredential {
  if (privateKey !== undefined) {
    if (typeof privateKey !== 'string') {
      throw new InvalidOptionsError('The Mesa constructor accepts only a private key string.');
    }
    return parsePrivateKey(privateKey);
  }

  const environmentPrivateKey = getEnvVar(PRIVATE_KEY_ENV_VAR);
  if (environmentPrivateKey !== undefined) {
    return parsePrivateKey(environmentPrivateKey);
  }
  throw new MissingPrivateKeyError('Missing private key. Pass a private key or set `MESA_PRIVATE_KEY`.');
}

type PrivateKeyAuthors = readonly [Author, ...Author[]];

export interface MesaOptions {
  /** Organization-root Ed25519 private key used to sign request and filesystem access tokens locally. */
  privateKey?: string;
  apiUrl?: string;
  fetch?: typeof globalThis.fetch;
  userAgent?: string;
  webhookSecret?: string;
}

/** Non-token options accepted by {@link FilesystemDefinition.mount}. */
export interface FsMountRuntimeOptions {
  cache?: {
    diskCache?: { path: string; maxSizeBytes?: number };
  };
  telemetry?: TelemetryConfig;
}

/**
 * A layout definition bundle produced by calling `mesa.fs({ layout, ttl })`:
 * the snapshotted layout plus its scoped operations. The definition is the only
 * way to mount a layout or mint its token; its `ttl` is the lifetime of
 * every token minted from it, whether by `token()` or under the hood by
 * `mount()`.
 */
export interface FilesystemDefinition {
  /** An independent plain-object snapshot of the validated layout. */
  layout(): Layout;
  /**
   * Mount this layout as the complete namespace, accepting the non-token
   * mount options (`cache`, `telemetry`). The mount's token lifetime is the
   * definition's `ttl`.
   */
  mount(options?: FsMountRuntimeOptions): Promise<MesaFileSystem>;
  /**
   * Mint the layout-scoped, least-privilege access token. Repository names map
   * to `read-repo` or `write-repo` according to each declaration's mode, and
   * the token uses the definition's `ttl`.
   */
  token(): Promise<AccessToken>;
}

export type AccessToken = {
  token: string;
  /** Exact expiry as an ISO 8601 string. */
  expires_at: string;
};

type FsLayoutOptions = {
  layout: Layout;
  /**
   * Lifetime of every token the definition mints, in seconds. Defaults to
   * 15 minutes and allows up to four hours. There is no refresh; once the
   * token expires, the mount fails closed.
   */
  ttl?: number;
  authors: PrivateKeyAuthors;
};
type RuntimeFsLayoutOptions = {
  layout: Layout;
  ttl?: number;
  authors?: PrivateKeyAuthors;
};

export class Mesa {
  readonly apiUrl: string;
  readonly org: ApiResources['org'];
  readonly repos: ApiResources['repos'];
  readonly content: ApiResources['content'];
  readonly bookmarks: ApiResources['bookmarks'];
  readonly changes: ApiResources['changes'];
  readonly diffs: ApiResources['diffs'];
  readonly webhookTargets: ApiResources['webhookTargets'];
  readonly webhooks: ApiResources['webhooks'];
  readonly fs: (options: FsLayoutOptions) => FilesystemDefinition;

  private readonly credential: PrivateKeyCredential;
  private readonly restClient: RestClient;
  private cachedWhoAmI: WhoamiResponse | null = null;

  constructor(options: MesaOptions = {}) {
    if (options === null || typeof options !== 'object' || Array.isArray(options)) {
      throw new InvalidOptionsError('The Mesa constructor accepts an options object.');
    }
    if ('auth' in options || 'accessToken' in options) {
      throw new InvalidOptionsError('The Mesa constructor accepts only `privateKey`; access tokens are not supported.');
    }
    if ((options as MesaOptions & { authors?: unknown }).authors !== undefined) {
      throw new InvalidOptionsError(
        'The `authors` option is not accepted by the Mesa constructor. Pass authors to the operation instead.'
      );
    }
    const credential = resolvePrivateKey(options.privateKey);
    this.credential = credential;

    this.apiUrl = normalizeUrl(options.apiUrl?.trim() || DEFAULT_API_URL);
    if (options.userAgent && looksLikePrivateKey(options.userAgent)) {
      throw new InvalidOptionsError('User-agent metadata must not contain Mesa private key material.');
    }
    this.restClient = createRestClient({
      accessToken: () => signAutomaticPrivateKeyAccessToken({ privateKey: credential, admin: true }).token,
      apiUrl: this.apiUrl,
      fetch: options.fetch,
      userAgent: options.userAgent,
    });

    const resources = createApiResources({
      restClient: this.restClient,
      orgSlug: credential.org,
      webhookSecret: options.webhookSecret,
      requestAttribution: {
        sign: (authors) => signAutomaticPrivateKeyAccessToken({ privateKey: credential, authors, admin: true }).token,
      },
    });

    this.org = resources.org;
    this.repos = resources.repos;
    this.content = resources.content;
    this.bookmarks = resources.bookmarks;
    this.changes = resources.changes;
    this.diffs = resources.diffs;
    this.webhookTargets = resources.webhookTargets;
    this.webhooks = resources.webhooks;

    // Layout minting and mounting are reachable only through a definition;
    // the layout carries no organization, so the client's org scopes both.
    const mintLayoutToken = async (
      layout: Layout,
      ttl: number | undefined,
      authors: PrivateKeyAuthors
    ): Promise<AccessToken> => {
      const { access } = this.deriveLayoutTokenRequest(layout, 'mesa.fs({ layout }).token()');
      // The layout definition is the only public mint path; it signs through
      // the client's private signer and can never request `admin`.
      const signed = signAutomaticPrivateKeyAccessToken({
        privateKey: this.credential,
        authors,
        access,
        ttlSeconds: ttl,
      });
      return { token: signed.token, expires_at: signed.expiresAt };
    };

    const mountLayout = async (
      definitionLayout: Layout,
      ttl: number | undefined,
      authors: PrivateKeyAuthors,
      options: FsMountRuntimeOptions
    ): Promise<MesaFileSystem> => {
      // Untyped callers migrating from the removed mesa.fs.mount() may still
      // pass definition options here; dropping them silently would shorten
      // the token lifetime (or skip authors) with no warning.
      if ('ttl' in options) {
        throw new InvalidOptionsError(
          'mount() does not accept `ttl`. Set it on the definition: mesa.fs({ layout, ttl })'
        );
      }
      if ('authors' in options) {
        throw new InvalidOptionsError(
          'mount() does not accept `authors`. Set them on the definition: mesa.fs({ layout, authors })'
        );
      }
      const { layout, org, access } = this.deriveLayoutTokenRequest(definitionLayout, 'mesa.fs({ layout }).mount()');
      // Mint a single access token for the mount's whole lifetime. There is
      // no refresh and no access-token replacement, so when the token expires the
      // mount expires with it.
      const token = signAutomaticPrivateKeyAccessToken({
        privateKey: this.credential,
        authors,
        access,
        ttlSeconds: ttl,
      }).token;
      return this.createFs(org, options, token, layout);
    };

    const defineLayout = ({ layout, ttl, authors: authorsInput }: RuntimeFsLayoutOptions): FilesystemDefinition => {
      if (layout === undefined) {
        // Catches untyped callers passing a bare path map instead of options.
        throw new InvalidOptionsError("mesa.fs() requires a 'layout'");
      }
      const authors = authorsInput === undefined ? undefined : normalizeAccessTokenAuthors(authorsInput);
      if (authors === undefined) {
        throw new InvalidOptionsError('Private-key layout definitions require a nonempty `authors` option.');
      }
      // Built eagerly, so an invalid record fails here at definition time.
      // Pre-typed `Record<string, Repo | Repo[]>` values still assign to
      // `Layout`, so keep the runtime absolute-path check for them.
      for (const path of Object.keys(layout)) {
        if (!path.startsWith('/')) {
          throw new Error(`mesa.fs(): top-level path '${path}' must be absolute`);
        }
      }
      const layoutJson = JSON.stringify(layout);
      MesaFileSystem.validateLayout(layoutJson);
      const prepared = JSON.parse(layoutJson) as Layout;
      const definition = {
        layout: () => JSON.parse(layoutJson) as Layout,
        mount: (options: FsMountRuntimeOptions = {}) => mountLayout(prepared, ttl, authors, options),
        token: () => mintLayoutToken(prepared, ttl, authors),
      };
      return definition;
    };
    this.fs = defineLayout;
  }

  private createFs(
    org: string,
    fsOptions: FsMountRuntimeOptions,
    accessToken: string,
    layout: Layout
  ): Promise<MesaFileSystem> {
    const config: MesaFileSystemConfig = {
      org,
      accessToken,
      cache: fsOptions.cache,
      apiBaseUrl: this.apiUrl,
      telemetry: fsOptions.telemetry,
      // Layout mounts present exactly the composed workspace, replacing the
      // canonical browse tree.
      repos: [],
      layout,
    };
    return MesaFileSystem.createAsync(config);
  }

  /**
   * Derive the least-privilege token request for a layout: its repositories
   * collected from every declaration and scoped by name (signed offline, no
   * repository lookup), plus the narrowest access each declared mode allows.
   * Shared by the definition's `mount()` and `token()` so the derivation
   * exists exactly once.
   */
  private deriveLayoutTokenRequest(
    layout: Layout,
    caller: string
  ): { layout: Layout; org: string; access: Record<string, 'read-repo' | 'write-repo'> } {
    // The layout carries no organization of its own; the client's
    // organization scopes the token.
    const org = this.org.slug;
    const declarations: Repo[] = [];
    const visit = (entries: Record<string, Repo | Repo[]>): void => {
      for (const entry of Object.values(entries)) {
        for (const declaration of Array.isArray(entry) ? entry : [entry]) {
          declarations.push(declaration);
          if (declaration.subPaths) visit(declaration.subPaths);
        }
      }
    };
    // The layout is a pure path map: every key is a declaration entry.
    visit(layout as Record<string, Repo | Repo[]>);
    if (declarations.length === 0) {
      throw new InvalidOptionsError(`${caller} requires at least one layout repository`);
    }
    // Repository names are user-controlled. A null prototype keeps a valid
    // repository named `__proto__` from invoking Object's legacy setter.
    const access = Object.create(null) as Record<string, 'read-repo' | 'write-repo'>;
    for (const declaration of declarations) {
      if (declaration.name === '*') {
        throw new InvalidOptionsError(`${caller} requires exact repository names`);
      }
      const level = declaration.mode === 'rw' ? 'write-repo' : 'read-repo';
      if (level === 'write-repo' || access[declaration.name] === undefined) {
        access[declaration.name] = level;
      }
    }
    return { layout, org, access };
  }

  async whoami(): Promise<WhoamiResponse> {
    if (this.cachedWhoAmI) {
      return this.cachedWhoAmI;
    }

    try {
      return await this.fetchWhoAmI();
    } catch (error) {
      throw new OrgResolutionError('Unable to resolve caller identity from /whoami', { cause: error });
    }
  }

  private async fetchWhoAmI(): Promise<WhoamiResponse> {
    const whoami = await this.restClient.whoami();
    this.cachedWhoAmI = whoami;
    return whoami;
  }
}

File: python/src/mesa_sdk/_mesa.py
from __future__ import annotations

import os
from typing import TYPE_CHECKING, Any
from urllib.parse import urlparse

from mesa_rest.api.org import whoami as whoami_api

from mesa_sdk._signing_key import (
    NormalizedSigningKeyAuthors,
    PrivateKeyCredential,
    SignedAccessToken,
    SigningKeyAccess,
    looks_like_private_key,
    parse_private_key,
    sign_automatic_private_key_access_token,
)
from mesa_sdk._client import BearerAuth, create_client, unwrap
from mesa_sdk._fs import FsNamespace
from mesa_sdk._resources import (
    Bookmarks,
    Changes,
    Content,
    Diffs,
    OrgResource,
    Repos,
    RequestAttribution,
    WebhookTargets,
)
from mesa_sdk._webhooks import Webhooks
from mesa_sdk.errors import (
    InvalidApiUrlError,
    InvalidOptionsError,
    MissingCredentialError,
)

if TYPE_CHECKING:
    from mesa_rest.client import AuthenticatedClient
    from mesa_rest.models.whoami_response_200 import WhoamiResponse200

DEFAULT_API_URL = "https://api.mesa.dev/v1"
PRIVATE_KEY_ENV_VAR = "MESA_PRIVATE_KEY"


def _resolve_private_key(private_key: str | None) -> PrivateKeyCredential:
    """Resolve an explicit or environment private key."""
    if private_key is not None:
        return parse_private_key(private_key)
    environment_private_key = os.environ.get(PRIVATE_KEY_ENV_VAR, "").strip()
    if environment_private_key:
        return parse_private_key(environment_private_key)

    raise MissingCredentialError(
        "Missing credential. Pass `private_key` or set "
        f"`{PRIVATE_KEY_ENV_VAR}` in your environment."
    )


def _normalize_url(url: str) -> str:
    if looks_like_private_key(url):
        raise InvalidApiUrlError("Mesa private keys cannot be used as URLs")
    try:
        parsed = urlparse(url.strip())
    except ValueError as exc:
        raise InvalidApiUrlError(url) from exc

    if parsed.scheme not in ("https", "http"):
        raise InvalidApiUrlError(url)

    return url.strip().rstrip("/")


class Mesa:
    """Async Mesa SDK client.

    Usage::

        async with Mesa(private_key="mesa_private_key_acme_...") as mesa:
            repos = await mesa.repos.list()
    """

    api_url: str

    repos: Repos
    bookmarks: Bookmarks
    changes: Changes
    content: Content
    diffs: Diffs
    org: OrgResource
    webhook_targets: WebhookTargets
    webhooks: Webhooks
    _credential: PrivateKeyCredential
    _client: AuthenticatedClient
    _cached_whoami: WhoamiResponse200 | None

    def __init__(
        self,
        *,
        private_key: str | None = None,
        api_url: str = DEFAULT_API_URL,
        user_agent: str | None = None,
        webhook_secret: str | None = None,
    ) -> None:
        """Construct a Mesa client.

        :param private_key: Organization-bound Ed25519 private key used to
            sign fresh request and filesystem credentials locally. Falls back
            to the ``MESA_PRIVATE_KEY`` environment variable.
        :param api_url: Override the default API endpoint.
        :param user_agent: Custom ``User-Agent`` header.
        :param webhook_secret: Secret used by ``mesa.webhooks.receive(...)``.

        :raises MissingCredentialError: If no credential is provided directly
            or through ``MESA_PRIVATE_KEY``.
        """
        private_key_credential = _resolve_private_key(private_key)
        self._credential = private_key_credential

        self.api_url = _normalize_url(api_url)
        if user_agent and looks_like_private_key(user_agent):
            raise InvalidOptionsError(
                "User-agent metadata must not contain Mesa private key material."
            )
        self._cached_whoami = None

        credential_org = private_key_credential.org
        bearer_auth = BearerAuth(
            lambda: (
                sign_automatic_private_key_access_token(
                    private_key=private_key_credential,
                    admin=True,
                ).token
            )
        )
        request_attribution = RequestAttribution(
            sign=lambda authors: (
                sign_automatic_private_key_access_token(
                    private_key=private_key_credential,
                    authors=authors,
                    admin=True,
                ).token
            ),
            auth=bearer_auth,
        )

        self._client = create_client(
            credential=bearer_auth,
            api_url=self.api_url,
            user_agent=user_agent,
        )

        self.repos = Repos(self._client, credential_org)
        self.bookmarks = Bookmarks(self._client, credential_org, request_attribution)
        self.changes = Changes(self._client, credential_org, request_attribution)
        self.content = Content(self._client, credential_org)
        self.diffs = Diffs(self._client, credential_org)
        self.org = OrgResource(self._client, credential_org)
        self.webhook_targets = WebhookTargets(self._client, credential_org)
        self.webhooks = Webhooks(webhook_secret)

    async def _create_mount_token(
        self,
        *,
        access: SigningKeyAccess,
        authors: NormalizedSigningKeyAuthors,
        ttl_seconds: int | None,
    ) -> str:
        """Return the bearer credential for a filesystem mount."""
        return self._sign_layout_token(
            access=access,
            authors=authors,
            ttl_seconds=ttl_seconds,
        ).token

    def _sign_layout_token(
        self,
        *,
        access: SigningKeyAccess,
        authors: NormalizedSigningKeyAuthors,
        ttl_seconds: int | None,
    ) -> SignedAccessToken:
        """Sign a private-key filesystem layout credential."""
        return sign_automatic_private_key_access_token(
            private_key=self._credential,
            authors=authors,
            access=access,
            ttl_seconds=ttl_seconds,
        )

    @property
    def fs(self) -> FsNamespace:
        """Filesystem namespace for mounting repos as a virtual filesystem.

        Use :meth:`mesa.fs(layout=...).mount() <mesa_sdk.FilesystemDefinition.mount>`
        to open a :class:`MesaFileSystem`.
        """
        if not hasattr(self, "_fs"):
            self._fs = FsNamespace(self)
        return self._fs

    async def whoami(self) -> WhoamiResponse200:
        """Return the identity tied to the credential. Cached after first call."""
        if self._cached_whoami is not None:
            return self._cached_whoami

        resp = await whoami_api.asyncio_detailed(client=self._client)
        result = unwrap(resp)
        self._cached_whoami = result
        return result

    async def __aenter__(self) -> Mesa:
        await self._client.__aenter__()
        return self

    async def __aexit__(self, *args: Any) -> None:
        await self._client.__aexit__(*args)
